### PR TITLE
FEC-10340 Multi-codec support for Audio/Video

### PR DESCRIPTION
**Pull request by GouravSna**
These APIs are part of `PlayerSettings`,

- `setPreferredVideoCodecSettings`: Set Preferred codec for video track. If the video has multiple codes (Ex: HEVC, AVC, AV1, VP9, VP8). 
`player.getSettings().setPreferredVideoCodecSettings(new VideoCodecSettings().setCodecPriorityList(Collections.singletonList(PKVideoCodec.HEVC)));
`

- `setPreferredAudioCodecSettings`: Set Preferred codec for audio track. If the video has multiple codes (Ex: AAC, AC3, E_AC3, OPUS).
`player.getSettings().setPreferredAudioCodecSettings(new AudioCodecSettings().setCodecPriorityList(Collections.singletonList(PKAudioCodec.AAC)));
`

	In both the above preferred codec Settings, you can pass your custom preferred list Singleton. There are other setters as well for Codec Settings,
	
	* `VideoCodecSettings` 
	
		`allowSoftwareDecoder` to allow the player to user software decoder if hardware decoder is not at all available, by default it is `false`.
	
		`allowMixedCodecAdaptiveness` if it is true then it will override your preferred codec list. It will add all the video codec in the tracks. If media has multiple codec available for video.
		
	* `AudioCodecSettings`
	
		`allowMixedCodecs` if it is true then it will override your preferred codec list. It will add all the audio codec in the tracks. If media has multiple codec available for audio.

### CODE REVIEW

**Review by giladna — APPROVED** [review]

File: playkit/src/main/java/com/kaltura/playkit/player/AudioCodecSettings.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
+package com.kaltura.playkit.player;
+
+import com.kaltura.playkit.PKAudioCodec;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class AudioCodecSettings {
+    private List<PKAudioCodec> audioCodecPriorityList = new ArrayList<>();
+    private boolean allowAudioMixedMimeTypes = false;
```
Comment (noamtamim) [inline playkit/src/main/java/com/kaltura/playkit/player/AudioCodecSettings.java]: Remove "audio" from field names
Change "MimeTypes" to "Codecs"

File: playkit/src/main/java/com/kaltura/playkit/player/VideoCodecSettings.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
+package com.kaltura.playkit.player;
+
+import com.kaltura.playkit.PKVideoCodec;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class VideoCodecSettings {
```
Comment (noamtamim) [inline playkit/src/main/java/com/kaltura/playkit/player/VideoCodecSettings.java:8]: Remove "video" from all fields and functions

File: playkit/src/main/java/com/kaltura/playkit/player/VideoCodecSettings.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    private List<PKVideoCodec> videoCodecPriorityList = new ArrayList<>();
+    private boolean allowSoftwareDecoder = false;
+    private boolean allowVideoMixedMimeTypeAdaptiveness = false;
```
Comment (noamtamim) [inline playkit/src/main/java/com/kaltura/playkit/player/VideoCodecSettings.java]: Rename "MimeType" to "Codec"

File: playkit/src/main/java/com/kaltura/playkit/player/TrackSelectionHelper.java [hunk 4]
```diff
@@ -254,10 +300,92 @@ private String getLanguageFromFormat(Format format) {
...
+            return  videoTracksCodecsMap.get(PKVideoCodec.AV1);
+        } else if (videoTracksCodecsMap.containsKey(PKVideoCodec.VP9)) {
+            return  videoTracksCodecsMap.get(PKVideoCodec.VP9);
+        } else if (videoTracksCodecsMap.containsKey(PKVideoCodec.VP8)) {
+            return  videoTracksCodecsMap.get(PKVideoCodec.VP8);
+        } else if (videoTracksCodecsMap.containsKey(PKVideoCodec.AVC)) {
+            return  videoTracksCodecsMap.get(PKVideoCodec.AVC);
+        } else {
+            return videoTracks;
+        }
```
Comment (noamtamim) [inline playkit/src/main/java/com/kaltura/playkit/player/TrackSelectionHelper.java]: For loop - create local array with all codecs by priority.

File: playkit/src/main/java/com/kaltura/playkit/player/TrackSelectionHelper.java [hunk 5]
```diff
@@ -279,6 +407,21 @@ private String getLanguageFromFormat(Format format) {
...
+
+        if (audioTracksCodecsMap.containsKey(PKAudioCodec.E_AC3)) {
+            return  new ArrayList<>(audioTracksCodecsMap.get(PKAudioCodec.E_AC3));
+        } else if (audioTracksCodecsMap.containsKey(PKAudioCodec.AC3)) {
+            return  new ArrayList<>(audioTracksCodecsMap.get(PKAudioCodec.AC3));
+        } else if (audioTracksCodecsMap.containsKey(PKAudioCodec.OPUS)) {
+            return  new ArrayList<>(audioTracksCodecsMap.get(PKAudioCodec.OPUS));
+        } else if (audioTracksCodecsMap.containsKey(PKAudioCodec.AAC)) {
+            return  new ArrayList<>(audioTracksCodecsMap.get(PKAudioCodec.AAC));
+        }
```
Comment (noamtamim) [inline playkit/src/main/java/com/kaltura/playkit/player/TrackSelectionHelper.java]: For loop like in video.

**Review by noamtamim — CHANGES_REQUESTED** [review]

File: playkit/src/main/java/com/kaltura/playkit/player/AudioCodecSettings.java (hunk 1 above)
Comment (GouravSna) [reply playkit/src/main/java/com/kaltura/playkit/player/AudioCodecSettings.java]: Refactored the name

File: playkit/src/main/java/com/kaltura/playkit/player/VideoCodecSettings.java (hunk 3 above)
Comment (GouravSna) [reply playkit/src/main/java/com/kaltura/playkit/player/VideoCodecSettings.java]: Refactored the name

File: playkit/src/main/java/com/kaltura/playkit/player/TrackSelectionHelper.java (hunk 5 above)
Comment (GouravSna) [reply playkit/src/main/java/com/kaltura/playkit/player/TrackSelectionHelper.java]: Converted to forloop

File: playkit/src/main/java/com/kaltura/playkit/player/TrackSelectionHelper.java (hunk 4 above)
Comment (GouravSna) [reply playkit/src/main/java/com/kaltura/playkit/player/TrackSelectionHelper.java]: Converted to forloop

File: playkit/src/main/java/com/kaltura/playkit/player/VideoCodecSettings.java (hunk 2 above)
Comment (GouravSna) [reply playkit/src/main/java/com/kaltura/playkit/player/VideoCodecSettings.java:8]: done

**Review by giladna — APPROVED** [review]

**Review by noamtamim — APPROVED** [review]

**Pull request merged**